File: mx_peer_discovery_py/mx_peer_discovery/mdns.py
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Multicast DNS service discovery.

Thin wrapper around the :mod:`zeroconf` library that provides a single
peer-oriented abstraction: register our own service once, continuously
browse for peers of the same service type, and emit a callback for each
resolved peer.

Default service type: ``_mx-peer._tcp.local.`` (note the trailing dot -
zeroconf requires it).

This implementation uses standard RFC 6763 SRV+A records for host and
port, and TXT purely for user metadata, matching the companion Rust
implementation (which wraps :crate:`mdns-sd`). The two halves interoperate
on the same LAN.
"""

import asyncio
import logging
import random
import socket
import string
from collections.abc import Callable
from dataclasses import dataclass, field

from zeroconf import ServiceInfo, ServiceStateChange
from zeroconf.asyncio import AsyncServiceBrowser, AsyncServiceInfo, AsyncZeroconf

log = logging.getLogger(__name__)

DEFAULT_SERVICE_TYPE = "_mx-peer._tcp.local."

OnResolved = Callable[[str, int, list[str], dict[str, str]], None]


@dataclass
class Config:
    """Configuration for :class:`MdnsDiscovery`.

    Mirrors the Rust ``Config`` struct field-for-field so cross-language
    callers describe their intent the same way.
    """

    hostname: str
    """Hostname advertised in the SRV record (typically ``"<instance>.local."``)."""

    ip: str
    """Single IP address advertised in the A record."""

    port: int
    """Port advertised in the SRV record."""

    on_resolved: OnResolved
    """Invoked for each resolved peer (excluding ourselves).

    Arguments: ``(instance_name, port, addresses, txt_properties)``.
    """

    txt: dict[str, str] = field(default_factory=dict)
    """Free-form metadata advertised in TXT records. Keys and values are
    limited to 255 bytes combined by the mDNS spec."""

    service_type: str = DEFAULT_SERVICE_TYPE
    """Service type (must end with ``.local.``)."""

    instance_name: str | None = None
    """Instance name (the label prefix on the service-type FQDN). Defaults
    to a random alphanumeric string when ``None``."""


class MdnsDiscovery:
    """A live mDNS presence: advertises one service and browses for peers.

    Call :meth:`start` to register and begin discovery, :meth:`stop` to
    shut down. Instances are not reusable across stop/start cycles.
    """

    def __init__(self, config: Config) -> None:
        self._config = config
        self._instance_name = config.instance_name or _random_instance_name()
        self._fullname = f"{self._instance_name}.{config.service_type}"
        self._zc: AsyncZeroconf | None = None
        self._browser: AsyncServiceBrowser | None = None

    @property
    def instance_name(self) -> str:
        """The resolved instance label (random if not explicitly provided)."""
        return self._instance_name

    async def start(self) -> None:
        """Register our service and start browsing for peers."""
        if self._zc is not None:
            raise RuntimeError("MdnsDiscovery already started")

        self._zc = AsyncZeroconf()

        info = ServiceInfo(
            type_=self._config.service_type,
            name=self._fullname,
            addresses=[socket.inet_aton(self._config.ip)],
            port=self._config.port,
            properties=self._encoded_properties(),
            server=self._config.hostname,
        )
        await self._zc.async_register_service(info)

        self._browser = AsyncServiceBrowser(
            self._zc.zeroconf,
            self._config.service_type,
            handlers=[self._on_state_change],
        )

    async def stop(self) -> None:
        """Unregister our service and close the zeroconf handle."""
        if self._browser is not None:
            await self._browser.async_cancel()
            self._browser = None
        if self._zc is not None:
            await self._zc.async_close()
            self._zc = None

    def _encoded_properties(self) -> dict[bytes, bytes]:
        """Zeroconf wants bytes on the wire, so encode up front."""
        return {
            k.encode("utf-8"): v.encode("utf-8") for k, v in self._config.txt.items()
        }

    def _on_state_change(
        self,
        zeroconf,
        service_type: str,
        name: str,
        state_change: ServiceStateChange,
    ) -> None:
        """ServiceBrowser handler. Dispatches resolution on a background task."""
        if state_change != ServiceStateChange.Added:
            return
        if name == self._fullname:
            return  # self-filter

        asyncio.ensure_future(self._resolve_and_dispatch(service_type, name))

    async def _resolve_and_dispatch(self, service_type: str, name: str) -> None:
        """Resolve a discovered service and invoke the user callback."""
        if self._zc is None:
            return

        info = AsyncServiceInfo(service_type, name)
        resolved = await info.async_request(self._zc.zeroconf, timeout=3000)
        if not resolved:
            log.debug(f"mdns: resolution timed out for {name}")
            return

        addresses = [socket.inet_ntoa(addr) for addr in info.addresses]
        port = info.port or 0
        txt = _decode_properties(info.properties or {})
        instance_label = _instance_label_from_fullname(name, service_type) or name

        try:
            self._config.on_resolved(instance_label, port, addresses, txt)
        except Exception as e:
            log.warning(f"on_resolved callback raised: {e}")


def _decode_properties(props: dict[bytes, bytes | None]) -> dict[str, str]:
    """Decode zeroconf's bytes-keyed property map back into a string dict.

    Zero-length values and None-valued keys are preserved as empty strings.
    Non-UTF-8 bytes are replaced rather than raising: mDNS TXT values are
    untrusted input.
    """
    result: dict[str, str] = {}
    for raw_key, raw_val in props.items():
        key = raw_key.decode("utf-8", errors="replace")
        val = raw_val.decode("utf-8", errors="replace") if raw_val else ""
        result[key] = val
    return result


def _instance_label_from_fullname(fullname: str, service_type: str) -> str | None:
    """Strip the service-type suffix to recover the bare instance label.

    Returns ``None`` when ``fullname`` does not end with the service type.
    """
    suffix = f".{service_type}"
    if fullname.endswith(suffix):
        return fullname[: -len(suffix)]
    return None


def _random_instance_name() -> str:
    """Generate a 32-character lowercase alphanumeric instance name."""
    alphabet = string.ascii_lowercase + string.digits
    return "".join(random.choices(alphabet, k=32))

File: mx_peer_discovery/src/mdns.rs
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

//! Multicast DNS service discovery.
//!
//! Thin wrapper around the [`mdns-sd`](https://crates.io/crates/mdns-sd) crate
//! that provides a single peer-oriented abstraction: register our own service
//! once, continuously browse for peers of the same service type, and emit a
//! callback for each resolved peer.
//!
//! Default service type: `_mx-peer._tcp.local.` (note the trailing dot -
//! mdns-sd requires it).
//!
//! # Interop with the Python half
//!
//! The Python companion module encodes host and port inside TXT records
//! (keys `host` and `port`), a libp2p-historical convention. This Rust
//! implementation uses standard RFC 6763 SRV+A records for host and port,
//! and TXT purely for user metadata. Cross-language discovery between
//! Python and Rust peers therefore requires the Python side to be
//! updated to emit and parse SRV+A records. Within a single language,
//! discovery works out of the box.

use std::collections::HashMap;
use std::sync::Arc;

use mdns_sd::{ServiceDaemon, ServiceEvent, ServiceInfo};
use tokio::task::JoinHandle;
use tracing::{debug, warn};

/// Default service type for MX peers. The trailing dot is required by
/// [`mdns_sd`].
pub const DEFAULT_SERVICE_TYPE: &str = "_mx-peer._tcp.local.";

/// Callback invoked for each resolved peer.
///
/// Arguments: `(instance_name, port, addresses, txt_properties)`. The
/// `addresses` set contains every IP the peer advertises; callers pick
/// one (or iterate) according to their routing preferences.
pub type OnResolved =
    Arc<dyn Fn(String, u16, Vec<String>, HashMap<String, String>) + Send + Sync + 'static>;

/// A live mDNS presence: advertises one service and browses for peers.
///
/// Drop the value to stop the daemon (best-effort shutdown).
pub struct MdnsDiscovery {
    daemon: ServiceDaemon,
    _browser: JoinHandle<()>,
}

/// Configuration for [`MdnsDiscovery::start`].
///
/// All fields are required; there's no builder yet because the surface is
/// small enough to pass directly.
pub struct Config {
    /// Service type (must end with `.local.`). Defaults to
    /// [`DEFAULT_SERVICE_TYPE`] when `None`.
    pub service_type: Option<String>,
    /// Instance name (the label prefix on the service-type FQDN).
    /// Defaults to a random alphanumeric string when `None`.
    pub instance_name: Option<String>,
    /// Hostname to register in the SRV record (typically `"<instance>.local."`).
    pub hostname: String,
    /// IP addresses this peer advertises. Comma-separated string or a
    /// single IP literal: see mdns-sd's `AsIpAddrs` for supported forms.
    pub ip: String,
    /// Port this peer listens on.
    pub port: u16,
    /// TXT properties for metadata. The mDNS wire spec limits each
    /// `key=value` to 255 bytes.
    pub txt: HashMap<String, String>,
    /// Invoked for each resolved peer (excluding ourselves).
    pub on_resolved: OnResolved,
}

impl MdnsDiscovery {
    /// Start the daemon, register our service, and begin browsing for peers.
    ///
    /// The returned value must be kept alive for discovery to continue;
    /// dropping it shuts the daemon down.
    ///
    /// # Errors
    ///
    /// Returns [`mdns_sd::Error`] if the daemon cannot be created, the
    /// service info is invalid, registration fails, or browsing cannot
    /// start.
    pub fn start(config: Config) -> Result<Self, mdns_sd::Error> {
        let service_type = config
            .service_type
            .unwrap_or_else(|| DEFAULT_SERVICE_TYPE.to_owned());
        let instance_name = config.instance_name.unwrap_or_else(random_instance_name);

        let daemon = ServiceDaemon::new()?;

        let service_info = ServiceInfo::new(
            &service_type,
            &instance_name,
            &config.hostname,
            config.ip.as_str(),
            config.port,
            config.txt,
        )?;
        daemon.register(service_info)?;

        let receiver = daemon.browse(&service_type)?;
        let on_resolved = config.on_resolved;
        let self_instance = instance_name;
        let browser = tokio::spawn(async move {
            while let Ok(event) = receiver.recv_async().await {
                match event {
                    ServiceEvent::ServiceResolved(resolved) => {
                        let fullname = resolved.get_fullname();
                        if is_own_instance(fullname, &self_instance) {
                            continue;
                        }
                        let addresses: Vec<String> = resolved
                            .get_addresses()
                            .iter()
                            .map(|scoped| scoped.to_string())
                            .collect();
                        let port = resolved.get_port();
                        let txt: HashMap<String, String> = resolved
                            .get_properties()
                            .iter()
                            .map(|p| (p.key().to_owned(), p.val_str().to_owned()))
                            .collect();
                        let instance_label = instance_label_from_fullname(fullname, &service_type)
                            .unwrap_or_else(|| fullname.to_owned());
                        on_resolved(instance_label, port, addresses, txt);
                    }
                    ServiceEvent::ServiceRemoved(ty, fullname) => {
                        debug!("service removed: {fullname} (type {ty})");
                    }
                    other => {
                        debug!("mdns event: {other:?}");
                    }
                }
            }
            debug!("mdns browser channel closed");
        });

        Ok(Self {
            daemon,
            _browser: browser,
        })
    }
}

impl Drop for MdnsDiscovery {
    fn drop(&mut self) {
        if let Err(e) = self.daemon.shutdown() {
            warn!("mdns daemon shutdown failed: {e}");
        }
    }
}

/// Returns true when the `fullname` belongs to our own registered instance.
///
/// Fullnames are formatted `{instance}.{service_type}`. A naive
/// `starts_with` check on a bare instance label is enough because
/// mdns-sd ensures the separator between the label and the service type
/// is always a `.`.
fn is_own_instance(fullname: &str, instance: &str) -> bool {
    let prefix = format!("{instance}.");
    fullname.starts_with(&prefix)
}

/// Strip the service-type suffix to recover the bare instance label.
///
/// Returns `None` when the fullname does not end with the service type.
fn instance_label_from_fullname(fullname: &str, service_type: &str) -> Option<String> {
    let suffix = format!(".{service_type}");
    fullname.strip_suffix(suffix.as_str()).map(str::to_owned)
}

/// Generate a 32-character lowercase alphanumeric instance name.
///
/// Not cryptographically random: we only need uniqueness on a LAN, and the
/// SystemTime + process-id mix is good enough for that. No external `rand`
/// dep.
fn random_instance_name() -> String {
    const ALPHABET: &[u8] = b"abcdefghijklmnopqrstuvwxyz0123456789";
    const LEN: usize = 32;
    const ALPHABET_LEN: u64 = 36;

    let mut state: u64 = seed_state();
    let mut out = String::with_capacity(LEN);
    for _ in 0..LEN {
        state = splitmix64(state);
        let idx = (state.checked_rem(ALPHABET_LEN).unwrap_or(0)) as usize;
        let byte = *ALPHABET.get(idx).unwrap_or(&b'0');
        out.push(byte as char);
    }
    out
}

/// Seed the PRNG from wall-clock nanos XOR a decorrelated process id.
fn seed_state() -> u64 {
    let nanos = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_nanos() as u64)
        .unwrap_or(0);
    let pid = u64::from(std::process::id());
    nanos ^ pid.wrapping_mul(0x9E37_79B9_7F4A_7C15)
}

/// SplitMix64: one step of the SplitMix64 PRNG.
fn splitmix64(x: u64) -> u64 {
    let mut z = x.wrapping_add(0x9E37_79B9_7F4A_7C15);
    z = (z ^ (z >> 30)).wrapping_mul(0xBF58_476D_1CE4_E5B9);
    z = (z ^ (z >> 27)).wrapping_mul(0x94D0_49BB_1331_11EB);
    z ^ (z >> 31)
}

#[cfg(test)]
#[allow(clippy::expect_used)]
mod tests {
    use super::*;

    #[test]
    fn default_service_type_constant() {
        assert_eq!(DEFAULT_SERVICE_TYPE, "_mx-peer._tcp.local.");
    }

    #[test]
    fn random_instance_name_shape() {
        let name = random_instance_name();
        assert_eq!(name.len(), 32);
        assert!(
            name.chars()
                .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit()),
            "expected lowercase alphanumeric, got {name:?}",
        );
    }

    #[test]
    fn random_instance_name_varies() {
        // Two draws from a 36^32 space. Collision probability is vanishingly
        // small; if this ever trips, the RNG seed is broken.
        let a = random_instance_name();
        std::thread::sleep(std::time::Duration::from_nanos(1));
        let b = random_instance_name();
        assert_ne!(a, b, "two draws produced identical values");
    }

    #[test]
    fn is_own_instance_matches_prefix() {
        assert!(is_own_instance(
            "myinstance._mx-peer._tcp.local.",
            "myinstance"
        ));
    }

    #[test]
    fn is_own_instance_rejects_different() {
        assert!(!is_own_instance(
            "otherinstance._mx-peer._tcp.local.",
            "myinstance"
        ));
    }

    #[test]
    fn is_own_instance_rejects_partial_prefix() {
        // "myinstance" is a prefix of "myinstance2" but not with the dot.
        assert!(!is_own_instance(
            "myinstance2._mx-peer._tcp.local.",
            "myinstance",
        ));
    }

    #[test]
    fn instance_label_extracts_from_fullname() {
        let label =
            instance_label_from_fullname("myinstance._mx-peer._tcp.local.", "_mx-peer._tcp.local.");
        assert_eq!(label, Some("myinstance".to_owned()));
    }

    #[test]
    fn instance_label_returns_none_on_wrong_suffix() {
        let label =
            instance_label_from_fullname("myinstance._other._tcp.local.", "_mx-peer._tcp.local.");
        assert_eq!(label, None);
    }

    #[test]
    fn splitmix64_changes_state() {
        assert_ne!(splitmix64(0), 0);
        assert_ne!(splitmix64(1), splitmix64(2));
    }

    // No end-to-end test of MdnsDiscovery::start - it binds to 224.0.0.251
    // which may not be available in sandboxed CI environments. Wire-format
    // correctness is covered by mdns-sd's own test suite.
}
